File: segment_tree.py

## BEGIN SEG TREE CLASS ==========
class SegTree:

    def __init__(self, numElements: int, defaultValue, associativeFunct):
        self.size = 1
        self.default = defaultValue
        self.funct = associativeFunct

        while self.size < numElements:
            self.size = (self.size<<1)

        self.st = [self.default for _ in range(self.size<<1)]

    ## CLASS FUNCTIONALITY ##########
    @staticmethod
    def _left(x):
        return (x<<1) + 1

    @staticmethod
    def _right(x): 
        return (x<<1) + 2

    @staticmethod
    def _mid(l, r):
        return (r-l)//2 + l

    ## END CLASS FUNCTIONALITY ##########


    ## PRIVATE FUNCTIONALITY ##########
    def _build(self, elements, x, lx, rx):
        if rx-lx == 1:
            if lx < len(elements):
                self.st[x] = elements[lx]
            return
        mx = SegTree._mid(lx,rx)
        self._build(elements, SegTree._left(x), lx, mx)
        self._build(elements, SegTree._right(x), mx, rx)
        self.st[x] = self.funct(self.st[SegTree._left(x)], self.st[SegTree._right(x)])

    def _set(self, i, v, x, lx, rx):
        if rx-lx == 1:
            self.st[x] = v
            return
        mx = SegTree._mid(lx,rx)
        if i < mx:
            self._set(i, v, SegTree._left(x), lx, mx)
        else:
            self._set(i, v, SegTree._right(x), mx, rx)

        self.st[x] = self.funct(self.st[SegTree._left(x)], self.st[SegTree._right(x)])

    def _range_query(self, l, r, x, lx, rx):
        if rx <= l or r <= lx:
            return self.default
        elif l <= lx and rx <= r:
            return self.st[x]
        else:
            mx = SegTree._mid(lx,rx)
            res1 = self._range_query(l,r,SegTree._left(x),lx,mx)
            res2 = self._range_query(l,r,SegTree._right(x),mx,rx)
            return self.funct(res1, res2)

    def _print_debug_elements(self, x, lx, rx):
        if rx-lx == 1:
            return ','.join(str(self.st[x+i]) for i in range(0,self.size))
        else:
            return self._print_debug_elements(SegTree._left(x), lx, SegTree._mid(lx,rx))

    ## END PRIVATE FUNCTIONALITY ##########


    ## INTERFACE FUNCTIONALITY ##########

    def build(self, elements: list):
        self._build(elements, 0, 0, self.size)

    def set(self, i, v):
        self._set(i, v, 0, 0, self.size)

    def range_query(self, l, r):
        return self._range_query(l, r, 0, 0, self.size)

    def __repr__(self):
        header = f'SegTree: (size={self.size}, default={self.default})'
        data = self._print_debug_elements(0, 0, self.size)
        return header + '\n' + data + '\n'

    ## END INTERFACE FUNCTIONALITY ##########

## END SEG TREE CLASS ##########

File: segment_tree.cpp
#include <bits/stdc++.h>

template <typename T>
class SegTree {
using AssociativeFunc = std::function<T (T&,T&)>;
private:

	static int left(int x) { return (x<<1) + 1; }
	static int right(int x) { return (x<<1) + 2; }
	static int mid(int l, int r) { return (r-l)/2 + l; }

	int size;			// width of the segment tree
	T defValue;			// default value used in the seg tree
	AssociativeFunc f;	// associative function that combines two segments
	std::vector<T> st;	// Stores the seg tree

	void build(std::vector<T>& a, int x, int lx, int rx) {
		if (rx-lx == 1) {
			if (lx < a.size()) {
				st[x] = a[lx];
			}
			return;
		} else {
			int mx = SegTree::mid(lx,rx);
			build(a, SegTree::left(x), lx, mx);
			build(a, SegTree::right(x), mx, rx);

			st[x] = f(st[SegTree::left(x)], st[SegTree::right(x)]);
		}
	}

	void set(int i, T v, int x, int lx, int rx) {
		if (rx-lx == 1) {
			st[x] = v;
			return;
		}

		int mx = SegTree::mid(lx,rx);
		if (i < mx) {
			set(i, v, SegTree::left(x), lx, mx);
		} else {
			set(i, v, SegTree::right(x), mx, rx);
		}

		st[x] = f(st[SegTree::left(x)], st[SegTree::right(x)]);
	}

	T rangeQuery(int l, int r, int x, int lx, int rx) {
		if (rx <= l || r <= lx) {
			return defValue;
		} else if (l <= lx && rx <= r) {
			return st[x];
		} else {
			int mx = SegTree::mid(lx,rx);
			
			T res1 = rangeQuery(l, r, SegTree::left(x), lx, mx);
			T res2 = rangeQuery(l, r, SegTree::right(x), mx, rx);
			return f(res1, res2);
		}
	}

	void debugPrintElements(int x, int lx, int rx) {
		if (rx-lx == 1) {
			std::cout << st[x] << ',';
			return;
		}

		int mx = SegTree::mid(lx, rx);
		debugPrintElements(SegTree::left(x), lx, mx);
		debugPrintElements(SegTree::right(x), mx, rx);
	}

public:
	SegTree(int numElements, T defaultValue, AssociativeFunc funct) 
	:defValue{defaultValue}, f{funct} {
		size = 1;
		while (size < numElements) { size = (size<<1); }
		st.assign(size<<1, defaultValue);
	}

	void build(std::vector<T>& elements) {
		build(elements, 0, 0, size);
	}

	void set(int i, T v) {
		set(i, v, 0, 0, size);
	}

	T rangeQuery(int l, int r) {
		return rangeQuery(l, r, 0, 0, size);
	}

	void debugPrintElements() {
		debugPrintElements(0, 0, size);
		std::cout << std::endl;
	}

};


int main() {


	return 0;
}
